Update Environment Variables (#1748)
zeit/docs: components/build-step-mdx/build-step.mdx

  deployment.
</Note>

## Environment Variables

> Environment Variables are accessible during both **Build Step** and **Runtime**, and can be configured for **Production**, **Preview** and **Development** \*\*Environments individually.

Reviewer: ```suggestion
Environment Variables are accessible during both **Build Step** and **Runtime**, and can be configured for **Production**, **Preview**, and **Development** \*\*Environments individually.
```

Oxford Comma 😃

This allows you to include values that you don't want to share in your code or to change the behavior of your code depending on its Environment.

To declare an Environment Variable for your deployment, head to the **General** page of your **Project Settings** and locate the **Environment Variables** section.

> First, enter the **Name** of the Environment Variable. For example, if you are using Next.js and you create an Environment Variable named `SENTRY_KEY`, it will be available under `process.env.SENTRY_KEY` in your code.

Reviewer: I think this is not really true. The "First" step is to select Production/Preview/Development, right?

Author: Fixed!

You can choose between the following **Environments**:

| Environment     | Description                                                                                                                                                                                     |
| --------------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| **Production**  | When selected, the Environment Variable will be applied to your next Production deployment. To create a Production deployment, push a commit to the default branch or run `now --prod`.         |
| **Preview**     | The Environment Variable is applied to your next Preview deployment. Preview deployments are created when you push to a branch (for example, `my-new-feature`) or run `now`.                    |
| **Development** | The Environment Variable is for use when running your project locally, with `now dev` or your preferred development command. To download Development Environment Variables, run `now env pull`. |

> Finally, click "Add" to create the Environment Variable. Once added, the Environment Variable is applied to your next deployment. It can then be consumed at Build Time by the framework of your choice, or by a serverless function during Runtime.

Reviewer: ```suggestion
Finally, click "Add" to create the Environment Variable. Once added, the Environment Variable is applied to your subsequent deployments. It can then be consumed at Build Time by the framework of your choice, or by a serverless function during Runtime.
```

  deployment.
</Note>

## Environment Variables

> Environment Variables are accessible during both **Build Step** and **Runtime**, and can be configured for **Production**, **Preview**, and **Development** \*\*Environments individually.

Reviewer: ```suggestion
Environment Variables are accessible during both **Build Step** and **Runtime**, and can be configured for **Production**, **Preview**, and **Development** Environments individually.
```

I think the asterisk here were a mistake?

This allows you to include values that you don't want to share in your code or to change the behavior of your code depending on its Environment.

To declare an Environment Variable for your deployment, head to the **General** page of your **Project Settings** and locate the **Environment Variables** section.

> **SCREENSHOT_EMPTY_UI**

Reviewer: Do we need a screenshot?

| **Production**  | When selected, the Environment Variable will be applied to your next Production deployment. To create a Production deployment, push a commit to the default branch or run `now --prod`.         |
| **Preview**     | The Environment Variable is applied to your next Preview deployment. Preview deployments are created when you push to a branch (for example, `my-new-feature`) or run `now`.                    |
| **Development** | The Environment Variable is for use when running your project locally, with `now dev` or your preferred development command. To download Development Environment Variables, run `now env pull`. |

Enter the **Name** of the Environment Variable. For example, if you are using Node.js and you create an Environment Variable named `SENTRY_KEY`, it will be available under `process.env.SENTRY_KEY` in your code.

Then, enter the **Value** of the Environment Variable. The value is automatically encrypted and stored as a _Secret_ in our system when you click **Add**.

> **SCREENSHOT_USER_ENV**

Reviewer: Do we need another screenshot?

Once added, the Environment Variable is applied to your subsequent deployment.
It can then be consumed during **Build Step** by the framework of your choice, or by a serverless function at **Runtime**.

<Note>

>   The way you set Environment Variables can affect the way the{' '}
>   <Link href="/docs/v2/platform/deployments#deduplication">
>     deduplication mechanism
>   </Link>{' '}
>   is applied. If your set of Environment Variables vary between Production and
>   Preview, deduplication can no longer be applied between Production and Preview
>   Deployments.

Reviewer: ```suggestion
  If any Environment Variables differ between Production and
  Preview,
  <Link href="/docs/v2/platform/deployments#deduplication">
    deduplication
  </Link> cannot be performed.
  This means Production will always created a new deployment.
```

I think we can re-word this note a bit.

Author: The meaning is not the same. `This means Production will always created a new deployment.` is not true, deduplication is only bypassed between Production and Preview. Still, I agree with the idea, I'll simplify!

Author: Simplified in https://github.com/zeit/docs/pull/1748/commits/502b65b76cebb96aa1f23f5645925b96b7ae6fe2.

</Note>

### Development Environment Variables

Environment Variables created for the Development Environment can be downloaded into a local development using a the `now env pull` command provided by [Now CLI](/download):

<Snippet clipboard={false} dark text={`now env pull
Downloading Development Environment Variables for project my-lovely-project

> ✅ Created file .env [510ms]`} />

Reviewer: ```suggestion
✅ Created .env file [510ms]`} />
```

Updated per https://github.com/zeit/now/pull/4059

<Caption>Downloading Development Environment Variables with `now env pull`.</Caption>

Running the command will create a `.env` file in your project folder. The `.env` file can then be consumed by `now dev` or
your preferred development command.

Check out [the Now CLI documentation](/docs/now-cli) to learn more about `now env` commands.

### Integration Environment Variables

[Integrations](/integrations) can automatically add Environment Variables to your Project Settings.
In that case, the Integration Configuration that added it will be displayed:

> **SCREENSHOT_INTEGRATION_ENV**

Reviewer: Making note of another screenshot missing

<Note>
  Support for Integration Environment Variables was added on April 10th, 2020.
  Variables added before that date do not have the <Strong>Integration</Strong>{' '}
  indication.
</Note>

### System Environment Variables

ZEIT provides a set of Environment Variables that are optionally populated by the System, such as the URL of the deployment or the name of the Git branch deployed.

To apply a System Environment Variable to your deployment, enter its **Name**, leave the **Value** empty, and click **Add**.

**SCREENSHOT_SYSTEM_ENV**

The following System Environment Variables are available:

<Table>
  <TableHead>
    <TableRow>
      <TableHeadCell className="head-cell">Name</TableHeadCell>
      <TableHeadCell className="head-cell">Description</TableHeadCell>
    </TableRow>
  </TableHead>
  <TableBody>
    {SYSTEM_ENVS.map(systemEnv => (
      <TableRow key={systemEnv.name}>
        <TableRowCell>
          <InlineCode>{systemEnv.name}</InlineCode>
        </TableRowCell>
        <TableRowCell>{systemEnv.description}</TableRowCell>
      </TableRow>
    ))}
  </TableBody>
</Table>

### Reserved Environment Variables

> The <ProductName /> platform reserves the usage of some Environment Variable names by default. You can find a list of these names in the [limits documentation](/docs/v2/platform/limits/#reserved-variables).

Reviewer: This list of reserved env vars needs to be updated to include anything that begins with `NOW_`.

</Note>

## Environment Variables

Environment Variables are accessible during both **Build Step** and **Runtime**, and can be configured for **Production**, **Preview**, and **Development** Environments individually.

This allows you to include values that you don't want to share in your code or to change the behavior of your code depending on its Environment.

To declare an Environment Variable for your deployment, head to the **General** page of your **Project Settings** and locate the **Environment Variables** section.

<Image
  src={`${process.env.ASSETS}/docs/build-step/env-vars-empty.png`}
  width={1488 / 2}
  height={824 / 2}
  shadow
/>
<Caption>The Environment Variables settings.</Caption>

First, choose between the following **Environments**:

| Environment     | Description                                                                                                                                                                                     |
| --------------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |

> | **Production**  | When selected, the Environment Variable will be applied to your next Production deployment. To create a Production deployment, push a commit to the default branch or run `now --prod`.         |

Reviewer 2: ```suggestion
| **Production**  | When selected, the Environment Variable will be applied to your next Production Deployment. To create a Production Deployment, push a commit to the default branch or run `now --prod`.         |
```

> | **Preview**     | The Environment Variable is applied to your next Preview deployment. Preview deployments are created when you push to a branch (for example, `my-new-feature`) or run `now`.                    |

Reviewer 2: ```suggestion
| **Preview**     | The Environment Variable is applied to your next Preview Deployment. Preview Deployments are created when you push to a branch (for example, `my-new-feature`) or run `now`.                    |
```

| **Development** | The Environment Variable is for use when running your project locally, with `now dev` or your preferred development command. To download Development Environment Variables, run `now env pull`. |

Enter the **Name** of the Environment Variable. For example, if you are using Node.js and you create an Environment Variable named `SENTRY_KEY`, it will be available under `process.env.SENTRY_KEY` in your code.

Then, enter the **Value** of the Environment Variable. The value is automatically encrypted and stored as a _Secret_ in our system when you click **Add**.

<Image
  src={`${process.env.ASSETS}/docs/build-step/env-vars-user.png`}
  width={1488 / 2}
  height={632 / 2}
  shadow
/>
<Caption>Adding an Environment Variable in the Project Settings.</Caption>

Once added, the Environment Variable is applied to your subsequent deployment.

> It can then be consumed during **Build Step** by the framework of your choice, or by a serverless function at **Runtime**.

Reviewer 3: ```suggestion
It can then be consumed during **Build Step** by the framework of your choice, or by a Serverless Function at **Runtime**.
```

<Note>

>   If any Environment Variable differ between Deployments,{' '}

Reviewer 3: ```suggestion
  If any Environment Variable differs between Deployments,{' '}
```

>   If any Environment Variable differs between Deployments,{' '}

Reviewer 3: ```suggestion
  If any Environment Variables differ between Deployments,{' '}
```

  <Link href="/docs/v2/platform/deployments#deduplication">deduplication</Link>{' '}
  will always be bypassed. For example, this means that merging to the default
  branch will no longer result in an instant deployment.
</Note>

### Development Environment Variables

> Environment Variables created for the Development Environment can be downloaded into a local development using a the `now env pull` command provided by [Now CLI](/download):

Reviewer 2: ```suggestion
Environment Variables created for the Development Environment can be downloaded into a local development setup using a the `now env pull` command provided by [Now CLI](/download):
```

> Environment Variables created for the Development Environment can be downloaded into a local development setup using a the `now env pull` command provided by [Now CLI](/download):

Reviewer 3: ```suggestion
Environment Variables created for the Development Environment can be downloaded into a local development setup using the `now env pull` command provided by [Now CLI](/download):
```

<Snippet clipboard={false} dark text={`now env pull
Downloading Development Environment Variables for project my-lovely-project
✅ Created .env file [510ms]`} />
<Caption>Downloading Development Environment Variables with `now env pull`.</Caption>

Running the command will create a `.env` file in your project folder. The `.env` file can then be consumed by `now dev` or
your preferred development command.

> Check out [the Now CLI documentation](/docs/now-cli#commands/env) to learn more about `now env` commands.

Reviewer 2: ```suggestion
Check out the [Now CLI reference](/docs/now-cli#commands/env) to learn more about `now env` commands.
```

### Integration Environment Variables

[Integrations](/integrations) can automatically add Environment Variables to your Project Settings.
In that case, the Integration Configuration that added it will be displayed:

<Image
  src={`${process.env.ASSETS}/docs/build-step/env-vars-integration.png`}
  width={1488 / 2}
  height={728 / 2}
  shadow
/>
<Caption>Adding an Integration Environment Variable in the Project Settings.</Caption>

<Note>
  Support for Integration Environment Variables was added on April 10th, 2020.
  Variables added before that date do not have the <Strong>Integration</Strong>{' '}
  indication.
</Note>

### System Environment Variables

> ZEIT provides a set of Environment Variables that are optionally populated by the System, such as the URL of the deployment or the name of the Git branch deployed.

Reviewer 2: ```suggestion
ZEIT provides a set of Environment Variables that are optionally populated by the System, such as the URL of the Deployment or the name of the Git branch deployed.
```

> To apply a System Environment Variable to your deployment, enter its **Name**, leave the **Value** empty, and click **Add**.

Reviewer 2: ```suggestion
To apply a System Environment Variable to your Deployment, enter its **Name**, leave the **Value** empty, and click **Add**.
```
